### PR TITLE
feat(plugin-codex): #1571 Codex host parity — PreCompact LCM flush + remote daemon

### LINKED ISSUE #1571: All host memory extensions must support the full featureset over remote/network (Codex extension is local-only)
## Principle

Every Remnic host integration should deliver the **full featureset** — automatic
recall injection, automatic observation, and compaction/LCM coordination —
against a **remote/central daemon over the network** (e.g. a shared Remnic daemon
on another host, reached over Tailscale), not just `localhost`.

> **Note (self-contained on purpose):** all detail is in this body, not only in
> the comments.

## Current state

| Host | Mechanism | Remote/central daemon? | Full featureset shipped? |
|------|-----------|------------------------|--------------------------|
| omp / Pi | `@remnic/plugin-pi` native extension (`context`/`turn_end`/`session_before_compact` → HTTP to `remnicDaemonUrl`) | ✅ | ✅ recall + observe + compaction |
| Claude Code | 4 session hooks (`engram-session-*.sh`) → `POST /engram/v1/recall` + `/observe` | ✅ | ✅ recall + observe |
| **Codex** | hooks (`~/.codex/hooks.json`) + local `memories_extensions` | ✅ (hooks) / ❌ (`memories_extensions` is local) | ❌ **not shipped/wired first-class** |

**Correction to an earlier assumption:** Codex is **not** MCP-only. Codex CLI
supports a full hook system — `SessionStart`, `UserPromptSubmit`, `Stop`,
`PreCompact`/`PostCompact`, `PostToolUse` (https://developers.openai.com/codex/hooks)
— so it *can* do automatic remote recall + observe + compaction, like Claude
Code. The `~/.codex/memories_extensions/remnic/instructions.md` path is a
**separate, complementary, local-only** ("No network") consolidation feature and
is not the remote path.

The gap is that Remnic ships first-class integrations for omp (`@remnic/plugin-pi`)
and Claude Code (the hook scripts), but **not** for Codex. In practice a working
recall script exists ad-hoc (`~/.codex/scripts/engram-session-recall.sh` → POST
`/engram/v1/recall`, returns `additionalContext`) but there is no observe or
compaction counterpart, and nothing is wired/shipped by Remnic itself.

## Codex hook I/O contract (verified)

`hooks.json` (or inline `[hooks]` in `config.toml`): `hooks.<Event>[] = {matcher, hooks:[{type:"command", command, timeout?}]}`.

- **SessionStart** — stdin has `session_id`, `cwd`, `source` (`startup|resume|clear|compact`), `transcript_path`. Return `{"continue":true,"hookSpecificOutput":{"hookEventName":"SessionStart","additionalContext":"…"}}` to inject recall. Matcher filters on `source`.
- **UserPromptSubmit** — stdin has `prompt`, `turn_id`, `cwd`. Same `additionalContext` output shape → per-turn recall.
- **Stop** (observe) — stdin has `last_assistant_message`, `turn_id`, `transcript_path`. **Must** return non-blocking JSON (`{}` / `{"continue":true}`); returning `{"decision":"block"}` makes Codex *continue* (loop). Observe should incrementally parse the Codex transcript (cursor) and POST `/engram/v1/observe` — this needs Codex's transcript JSONL schema (differs from Claude Code's; ref impl: Claude's `engram-session-store.sh`).
- **PreCompact** — stdin has `trigger` (`manual|auto`). POST `/engram/v1/lcm/compaction/flush`; return `{"continue":true}`.

## Blocker: hook trust

Codex records trust in `config.toml` as
`[hooks.state."<hooks.json path>:<event>:<group_idx>:<hook_idx>"]` with
`trusted_hash = "sha256:…"`. There is **no non-interactive trust CLI** — trust is
set via the interactive `/hooks` review, or `--dangerously-bypass-hook-trust`
per run (persistent security downgrade), or **managed/pre-trusted hooks**
(enterprise/MDM only). The hash keying is non-obvious (not `sha256(command)`).
So ad-hoc user hooks can't be trust-provisioned headlessly across a fleet.

## Ask

1. Ship a **first-class Codex integration** — e.g. `remnic connectors install codex`
   that installs recall/observe/precompact hook scripts and wires `~/.codex/hooks.json`
   against a configurable remote `remnicDaemonUrl` + token, **sharing the
   `@remnic/plugin-pi` client/transport** (don't re-implement recall/observe per host).
2. Handle trust as **managed/pre-trusted hooks** where possible, and otherwise
   document the one-time `/hooks` trust (or `--dangerously-bypass-hook-trust`).
3. Establish the invariant: **all host extensions support the full featureset
   including remote/network daemon connections.**

## Context

Found while wiring a homelab fleet to a central macstudio daemon over Tailscale
(codex-a/b/c, claude-a/b, 2 laptops, macstudio). omp got the full native
extension; Claude Code got its 4 hooks; Codex currently has MCP + (as an interim)
a `SessionStart` recall hook wired via `--dangerously-bypass-hook-trust`. Observe
+ compaction for Codex are intentionally deferred to this first-class work.

### PR DESCRIPTION
**Pull request by joshuaswarren**
Closes #1571.

## What

Bring the Codex adapter to full featureset parity with `@remnic/plugin-pi` and `@remnic/plugin-claude-code`. Every Codex host can now deliver **recall + observe + compaction coordination** against a **remote/central daemon over the network** (Tailscale/LAN/VPN), not just localhost.

### PreCompact hook (the missing piece)

- Wires `hooks.json` `PreCompact` (matcher `manual|auto`) to a new `pre-compact` handler that POSTs `/engram/v1/lcm/compaction/flush` before Codex compacts — mirroring `@remnic/plugin-pi`'s `session_before_compact` handler so the daemon drains the in-flight observe buffer into long-term memory before the transcript is summarized.
- **Always returns `continue:true`** — fail-open end to end. A flush failure (no token, HTTP 5xx, dead daemon) never blocks compaction. The upstream Codex contract notes `continue:false` stops the compact entirely, so this invariant is load-bearing.

### Remote / network daemon URL (the parity ask)

- Adds `REMNIC_DAEMON_URL` (and legacy `ENGRAM_DAEMON_URL` alias): a full base URL that lets a Codex host reach a shared central daemon. An explicit `http:`/`https:` scheme selects the transport; `https://` uses the TLS transport for daemons behind a reverse proxy. Same transport contract as `@remnic/plugin-pi`'s `remnicDaemonUrl`.
- `REMNIC_HOST`/`REMNIC_PORT` remain a backward-compat fallback (existing installs and the test harness keep working).
- Optional `REMNIC_NAMESPACE` attaches a namespace to the flush (parity with `config.namespace`).

### Docs + trust

README documents PreCompact, the remote daemon URL env vars, and the one-time `/hooks` trust story (managed/pre-trusted hooks for fleets via `requirements.toml` `managed_dir`, or `--dangerously-bypass-hook-trust` for spikes; non-managed hooks are skip-no-op until trusted, never a crash).

## Why

The issue identified that Remnic shipped first-class hook integrations for omp/Pi (`@remnic/plugin-pi`) and Claude Code (the 4 `engram-session-*.sh` scripts), but **not** for Codex — Codex had MCP + an ad-hoc SessionStart recall script, with no observe or compaction counterpart and no first-class wiring. This establishes the invariant the issue asks for: *all host extensions support the full featureset including remote/network daemon connections.*

## How verified

- **Upstream contract verified** against https://developers.openai.com/codex/hooks (AGENTS.md rule 5) before implementing: `PreCompact` added in v0.129.0; matcher filters `trigger` (`manual|auto`); `continue:false` stops compaction; `Stop` returns JSON on stdout with `decision:block` meaning continue.
- **Smoke test** (mock daemon): `pre-compact` POSTs the flush with `Bearer` auth + `{sessionKey}`, returns `{"continue":true}`; no-token → bare continue, no flush; `REMNIC_DAEMON_URL` overrides `REMNIC_PORT` (remote routing proven); daemon-down → still `continue:true`.
- **Test suite**: 8 new cases → 29/29 `node --test packages/plugin-codex/hooks/bin/remnic-codex-hook.test.cjs` green.
- **Preflight**: `npm run preflight:quick` → `[preflight] OK (quick)` (lint, check-types, check-config-contract, plugin:inspect, check-review-patterns, check-ratchets, check-docs-parity, all quick suites). `test:entity-hardening` correctly skipped — no core retrieval/intent/cache/config files touched.

## Test-harness hygiene (drive-by)

`runHook` now clears both token env vars when a test requests `no-token`. The parent shell's `OPENCLAW_ENGRAM_ACCESS_TOKEN` previously leaked through `...process.env` (the runner's `resolveToken()` falls through `OPENCLAW_REMNIC_ACCESS_TOKEN → OPENCLAW_ENGRAM_ACCESS_TOKEN`), making two pre-existing no-token tests fail on dev machines outside CI. Now hermetic.

## Scope

`packages/plugin-codex` only — no `@remnic/core` internals touched. The runner consumes the public HTTP surface (`/engram/v1/recall`, `/observe`, `/lcm/compaction/flush`, `/health`) the same way `@remnic/plugin-pi` does.

**Chokepoints used:** none applicable (plugin-package hook-script work; consumes the public `@remnic/core` HTTP surface, no core internals, no orchestrator/storage/intent/config edits).

+360 / −10 across 4 files + 1 changeset.

---

> [!NOTE]
> **Medium Risk**
> Changes session memory timing around compaction and cross-store routing via daemon URL/namespace; mistakes could lose transcript tails or write to the wrong namespace, but behavior is fail-open and heavily tested in plugin-codex only.
> 
> **Overview**
> Adds **Codex `PreCompact` parity** with `@remnic/plugin-pi`: `hooks.json` wires a new `pre-compact` handler that **drains any unobserved transcript tail** to `/engram/v1/observe`, then POSTs **`/engram/v1/lcm/compaction/flush`** before Codex summarizes. It always returns **`continue:true`** (fail-open); if the observe lock stays busy, the flush is **deferred** rather than racing a background worker.
> 
> **Remote daemon transport** routes all hook HTTP through **`REMNIC_DAEMON_URL`** (http/https, optional path prefix for reverse proxies). Invalid explicit URLs **disable** the daemon instead of falling back to localhost. **`withNamespace()`** applies `REMNIC_NAMESPACE` on recall, observe, and flush so namespaced installs stay consistent.
> 
> README covers remote setup and hook trust; integration tests expand for PreCompact, namespaces, daemon URL edge cases, and hermetic token env in `runHook`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b2ee6203b4b5838aea1a2100d0385e9828baa5b0. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs [hunk 1]
```diff
@@ -849,6 +895,37 @@ async function handleSessionEnd(input, token, log) {
...
+  if (!token) {
+    log(`skipping LCM flush: no token (session=${sessionId} trigger=${trigger})`);
+    return;
+  }
+  const body = { sessionKey: sessionId };
+  // Optional namespace override (parity with @remnic/plugin-pi's config.namespace).
+  // Unset → the daemon resolves the default namespace for this principal.
+  const ns = process.env.REMNIC_NAMESPACE || process.env.ENGRAM_NAMESPACE || "";
+  if (ns) body.namespace = ns;
+  const res = await httpPost("/engram/v1/lcm/compaction/flush", token, body, 20000);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs]: Fixed in ea063e89. `handlePreCompact` now drains the unobserved transcript tail to `/engram/v1/observe` **before** POSTing `/lcm/compaction/flush`, via a new `drainTranscriptTail` helper that:

- reads the session cursor and parses `parseTranscript(transcript_path).slice(lastCount)` — exactly the post-last-PostToolUse delta;
- acquires the session lock (so it cannot race the detached PostToolUse observe worker; if the lock is busy it skips and the worker drains the tail instead);
- advances the cursor (does not remove it) on a successful observe, since the session continues after compaction and the post-compact transcript must not re-observe this tail;
- retains the cursor on parse/observe failure so the next observe/compact retries.

Order is asserted by `pre-compact: drains the unobserved transcript tail to /observe BEFORE the LCM flush (#1571 review)`: `observeIdx < flushIdx`, the observe payload is exactly the 2-message pending tail, and the cursor advances to 4. A second test (`retains the cursor when the tail-drain observe fails`) proves cursor retention + fail-open `continue:true` when the observe 500s. 31/31 plugin-codex hook tests green; `preflight:quick` OK.

File: packages/plugin-codex/hooks/hooks.json [hunk 2]
```diff
@@ -51,6 +51,19 @@
...
+    ],
+    "PreCompact": [
+      {
+        "matcher": "manual|auto",
+        "hooks": [
+          {
+            "type": "command",
+            "command": "\"${PLUGIN_ROOT}/hooks/bin/remnic-codex-hook.sh\" pre-compact",
+            "commandWindows": "powershell -NoProfile -ExecutionPolicy Bypass -File \"${PLUGIN_ROOT}\\hooks\\bin\\remnic-codex-hook.ps1\" pre-compact",
+            "timeout": 20000
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/hooks.json]: Fixed in 4ca954f3. hooks.json PreCompact timeout bumped 20s → 90s to cover the full worst case: lock-wait budget (PRECOMPACT_LOCK_RETRIES, 15s) + drain /observe (30s) + LCM flush (20s) + overhead. `continue:true` is still emitted synchronously before any async work, so Codex never waits on the network for compaction to proceed — the timeout only governs how long the background coordination runs before being killed.

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs [hunk 3]
```diff
@@ -849,6 +895,109 @@ async function handleSessionEnd(input, token, log) {
...
+  }
+  const state = resolveState(sessionId, log);
+  if (!state) return false;
+  const { cursorFile, lockFile } = state;
+  // The lock guards against the detached post-tool observe worker racing this
+  // drain. If we can't acquire it (worker still running), skip — the worker
+  // will observe the tail, and the LCM flush still drains what's queued.
+  if (!acquireLock(lockFile, log)) {
+    log(`transcript tail drain skipped for ${sessionId}: lock busy`);
+    return false;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs]: Fixed in 4ca954f3. `drainTranscriptTail` now waits a bounded budget (`PRECOMPACT_LOCK_RETRIES`, default 150×100ms≈15s, `REMNIC_PRECOMPACT_LOCK_RETRIES`-tunable) for the detached observe worker to release the session lock, and returns `"busy"` if it outlasts the budget. `handlePreCompact` **skips the flush** on `"busy"` rather than racing ahead — the worker's in-flight `/observe` would otherwise miss this compaction, so we defer to the next PreCompact/Stop. `acquireLock` is parameterized (`retries` arg, default 50 unchanged for foreground hooks). Proven by `pre-compact: when the session lock stays held, the drain returns busy and the LCM flush is skipped` (holds the lock, asserts neither `/observe` nor `/lcm/compaction/flush` runs).

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs [hunk 4]
```diff
@@ -849,6 +895,109 @@ async function handleSessionEnd(input, token, log) {
...
+    } catch {
+      log(`transcript tail parse failed for ${sessionId}; cursor retained`);
+      return false;
+    }
+    if (newMessages.length === 0) return true;
+    log(`transcript tail drain: ${newMessages.length} new message(s) for ${sessionId}`);
+    const res = await httpPost(
+      "/engram/v1/observe",
+      token,
+      { sessionKey: sessionId, messages: newMessages },
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs]: Fixed in 4ca954f3. `drainTranscriptTail` now accepts a `namespace` arg and forwards `REMNIC_NAMESPACE`/`ENGRAM_NAMESPACE` on the tail `/observe` body, so a namespaced install queues the tail under the same key the LCM flush drains. `handlePreCompact` resolves the namespace once and passes it to both the drain and the flush. Proven by `pre-compact: REMNIC_NAMESPACE also scopes the tail-drain /observe to the same key` — asserts `observe.body.namespace === "team/fleet"` AND `flush.body.namespace === "team/fleet"` with a 2-message pending tail.

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs [hunk 5]
```diff
@@ -26,12 +26,42 @@ const fs = require("fs");
...
+// for TLS daemons (common for remote/private deployments behind a proxy).
+const DAEMON_URL = (function resolveDaemonUrl() {
+  if (RAW_DAEMON_URL) {
+    try {
+      const parsed = new URL(RAW_DAEMON_URL);
+      // Require an explicit scheme — a bare "macstudio:4318" parses as
+      // pathname-only and would silently route to localhost.
+      if (parsed.protocol === "http:" || parsed.protocol === "https:") return parsed;
+    } catch {
+      /* fall through to host/port */
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs]: Fixed in 4ca954f3. An explicit but malformed `REMNIC_DAEMON_URL` (missing scheme, e.g. `macstudio:4318`) no longer silently falls through to `REMNIC_HOST`/`REMNIC_PORT` — that would route recall/observe/flush to localhost and corrupt the wrong store. It now sets `DAEMON_URL = null` + a `DAEMON_CONFIG_ERROR`, and `httpPost`/`httpHealthy` fail open immediately (callers treat it like a dead daemon). A one-time stderr warning names the bad URL and how to fix it. Unset `REMNIC_DAEMON_URL` still uses HOST/PORT (backward compat preserved). Proven by `pre-compact: an invalid explicit REMNIC_DAEMON_URL disables the daemon instead of falling back to localhost` — asserts `calls.length === 0` against the localhost mock AND `continue:true` (fail-open).

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs [hunk 6]
```diff
@@ -849,6 +931,144 @@ async function handleSessionEnd(input, token, log) {
...
+    ns,
+  );
+  if (drainResult === "busy") {
+    log(`skipping LCM flush: tail drain lock busy (session=${sessionId} trigger=${trigger}); deferring to next cycle`);
+    return;
+  }
+  // 2. Ask the LCM layer to flush the (now-complete) observe buffer.
+  const body = { sessionKey: sessionId };
+  if (ns) body.namespace = ns;
+  const res = await httpPost("/engram/v1/lcm/compaction/flush", token, body, 20000);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs]: Fixed in ca2ead1e. Hardened the namespace defect class at a chokepoint: new `resolveNamespace()` + `withNamespace(body)` helpers, and EVERY observe/flush body now goes through `withNamespace()` — observeWorker (PostToolUse), handleSessionEnd final flush, drainTranscriptTail (PreCompact), and the PreCompact LCM flush. A namespaced Codex install now archives + flushes under one key end to end. drainTranscriptTail/handlePreCompact dropped their now-redundant per-call namespace handling. Proven by two new tests: `observe worker: REMNIC_NAMESPACE scopes the PostToolUse /observe` (asserts `observe.body.namespace === "fleet/alpha"`) and `session-end: REMNIC_NAMESPACE scopes the final-flush /observe` (asserts `observe.body.namespace === "fleet/beta"`).

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs [hunk 7]
```diff
@@ -849,6 +931,144 @@ async function handleSessionEnd(input, token, log) {
...
+    PRECOMPACT_LOCK_RETRIES,
+    ns,
+  );
+  if (drainResult === "busy") {
+    log(`skipping LCM flush: tail drain lock busy (session=${sessionId} trigger=${trigger}); deferring to next cycle`);
+    return;
+  }
+  // 2. Ask the LCM layer to flush the (now-complete) observe buffer.
+  const body = { sessionKey: sessionId };
+  if (ns) body.namespace = ns;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs]: Same defect class as the cursor thread — fixed together in ca2ead1e. All observe paths (PostToolUse worker + session-end + PreCompact tail drain) and the LCM flush now route through a single `withNamespace()` chokepoint reading `REMNIC_NAMESPACE`/`ENGRAM_NAMESPACE`, so a namespaced install flushes the same queue it archived to. See the worker + session-end namespace tests.

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.test.cjs [hunk 8]
```diff
@@ -57,8 +64,8 @@ function runHook(event, input, { port, home, env = {} } = {}) {
         REMNIC_HOST: "127.0.0.1",
         REMNIC_PORT: String(port),
         REMNIC_CODEX_MATERIALIZE: "0",
-        // Default to an env token unless a test overrides it.
-        OPENCLAW_REMNIC_ACCESS_TOKEN: env.token === null ? "" : env.token || "test-token",
+        OPENCLAW_REMNIC_ACCESS_TOKEN: noToken ? "" : env.token || "test-token",
+        OPENCLAW_ENGRAM_ACCESS_TOKEN: noToken ? "" : "",
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.test.cjs:68]: Fixed in ca2ead1e. `runHook` now sets `REMNIC_DAEMON_URL: ""` and `ENGRAM_DAEMON_URL: ""` in the base child env, so a developer shell with either set can no longer route tests away from the mock server. Tests that want a daemon URL override via `env.extra` (which spreads last), so the remote-routing + invalid-URL tests still pass.

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs [hunk 9]
```diff
@@ -26,28 +26,86 @@ const fs = require("fs");
...
 
+// PreCompact tail-drain lock-wait budget (100ms each → ~15s). Bounds how long
+// handlePreCompact waits for a detached PostToolUse observe worker (whose own
+// /observe can run up to 120s) to release the session lock before skipping the
+// drain+flush. 15s covers a typical observe round-trip on a healthy daemon; a
+// pathologically slow worker that outlasts this is left for the next cycle
+// rather than blocking compaction indefinitely. The hooks.json PreCompact
+// timeout accommodates this budget plus the drain observe + LCM flush.
+const PRECOMPACT_LOCK_RETRIES =
+  Number.parseInt(process.env.REMNIC_PRECOMPACT_LOCK_RETRIES || "150", 10) || 150;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs]: Fixed in 3cab73a3. Replaced `Number.parseInt(... || "150") || 150` with an explicit finite/non-negative check, so `REMNIC_PRECOMPACT_LOCK_RETRIES=0` is honored (immediate busy-skip) rather than coerced to 150. NaN/negative still falls back to 150. Proven by `pre-compact: REMNIC_PRECOMPACT_LOCK_RETRIES=0 is honored, not coerced to 150` (holds the lock, asserts flush skipped AND elapsed < 5s — a coerced 150 would impose ~15s).

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs [hunk 10]
```diff
@@ -201,10 +285,11 @@ function httpPost(urlPath, token, bodyObj, timeoutMs) {
...
       return;
     }
-    const req = http.request(
+    const req = transport.request(
       {
-        host: HOST,
-        port: PORT,
+        protocol: DAEMON_URL.protocol,
+        hostname: DAEMON_URL.hostname,
+        port: DAEMON_URL.port || (DAEMON_URL.protocol === "https:" ? 443 : 80),
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs:302]: Fixed in ea933589. Added `DAEMON_BASE_PATH` (= `DAEMON_URL.pathname` with trailing slashes stripped) and prepended it to every request path in httpPost/httpHealthy — so a reverse-proxy subpath mount like `http://gw/remnic` now routes to `/remnic/engram/v1/...`, matching @remnic/plugin-pi's daemon-URL + route concatenation. Proven by `session-start: a path-prefixed REMNIC_DAEMON_URL routes under the prefix` (asserts both recall and health hit `/remnic/engram/v1/...`).

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs [hunk 11]
```diff
@@ -201,10 +285,11 @@ function httpPost(urlPath, token, bodyObj, timeoutMs) {
       resolve({ ok: false, status: 0, body: "" });
       return;
     }
-    const req = http.request(
+    const req = transport.request(
       {
-        host: HOST,
-        port: PORT,
+        protocol: DAEMON_URL.protocol,
+        hostname: DAEMON_URL.hostname,
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs:301]: Not a bug — false positive, verified empirically. `new URL("http://[::1]:4318").hostname` returns the bracketed `"[::1]"`, and Node's `http.request({ hostname: "[::1]", port })` accepts that directly: a round-trip against a `::1` listener returns HTTP 200 with no ENOTFOUND. Node's net layer strips the IPv6 brackets internally, so recall/observe/flush already work for IPv6-only/Tailscale hosts. No code change needed.

File: packages/plugin-codex/hooks/hooks.json [hunk 12]
```diff
@@ -51,6 +51,19 @@
...
+    ],
+    "PreCompact": [
+      {
+        "matcher": "manual|auto",
+        "hooks": [
+          {
+            "type": "command",
+            "command": "\"${PLUGIN_ROOT}/hooks/bin/remnic-codex-hook.sh\" pre-compact",
+            "commandWindows": "powershell -NoProfile -ExecutionPolicy Bypass -File \"${PLUGIN_ROOT}\\hooks\\bin\\remnic-codex-hook.ps1\" pre-compact",
+            "timeout": 90000
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/hooks.json]: Fixed in ea933589. Verified against the official contract (developers.openai.com/codex/hooks: "`timeout` is in seconds"; default 600) — you're right, the prior 90000 meant 25h. Hardened the whole timeout class to seconds in one commit: SessionStart 45000→45, PostToolUse 10000→10, UserPromptSubmit 20000→20, Stop 30000→30, PreCompact 90000→90 (each prior ms value / 1000 = the intended ceiling). PreCompact at 90s covers the worst case (15s lock-wait + 30s drain observe + 20s flush + overhead).

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs [hunk 13]
```diff
@@ -492,8 +595,12 @@ function migrateTmpCursor(sessionId, cursorFile, log) {
...
 
 // mkdir-based mutex with stale-lock reaping (10 min). Returns true if acquired.
-function acquireLock(lockFile, log) {
-  for (let i = 0; i < 50; i++) {
+// `retries` bounds the synchronous wait (100ms each). The default (50 → ~5s)
+// suits foreground hooks; PreCompact's tail drain passes a larger budget so it
+// can outlast a detached PostToolUse observe worker holding the lock, ensuring
+// the subsequent flush sees the worker's queued messages instead of racing.
+function acquireLock(lockFile, log, retries = 50) {
+  for (let i = 0; i < retries; i++) {
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs]: Fixed in 25f00356. `acquireLock` now computes `attempts = Math.max(1, retries)`, so 0 retries means "try once, don't wait/retry" — a free lock is still taken (drain+flush proceed); only a busy lock skips immediately. Proven by `pre-compact: REMNIC_PRECOMPACT_LOCK_RETRIES=0 still takes a FREE lock` (no lock held, 0 retries → drain + flush both run). The existing held-lock test still passes (1 attempt fails → busy → flush skipped).

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs [hunk 14]
```diff
@@ -190,9 +250,43 @@ function resolveToken() {
...
+// archives + flushes under one key — without it, buffered observations land on
+// the default key while a namespaced flush drains a different (empty) queue, and
+// most in-session memory never flushes before compaction.
+function resolveNamespace() {
+  return process.env.REMNIC_NAMESPACE || process.env.ENGRAM_NAMESPACE || "";
+}
+function withNamespace(body) {
+  const ns = resolveNamespace();
+  if (ns) body.namespace = ns;
+  return body;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs:265]: Fixed in 25f00356. Extended the `withNamespace()` chokepoint to all 3 recall bodies (session-start full + minimal fallback + user-prompt-recall). Verified plugin-pi's `recall()` sets `namespace: this.config.namespace` on every recall body, so this is true parity. Proven by `session-start: REMNIC_NAMESPACE scopes the recall body` (asserts `recall.body.namespace === "fleet/gamma"`). Every HTTP body — recall, observe, flush — now routes through the single namespace chokepoint.

File: packages/plugin-codex/hooks/hooks.json [hunk 15]
```diff
@@ -47,7 +47,20 @@
             "type": "command",
             "command": "\"${PLUGIN_ROOT}/hooks/bin/remnic-codex-hook.sh\" session-end",
             "commandWindows": "powershell -NoProfile -ExecutionPolicy Bypass -File \"${PLUGIN_ROOT}\\hooks\\bin\\remnic-codex-hook.ps1\" session-end",
-            "timeout": 30000
+            "timeout": 30
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/hooks.json:50]: Won't-fix (factual correction + fail-open). The Stop hook `timeout` is **30 seconds**, not 90 (90 is PreCompact — see hooks.json line 50 vs 63). The observe+materialize chain can exceed 30s on a pathologically slow daemon, but `handleSessionEnd` emits `{continue:true}` as its FIRST statement (before any network/child work), so Codex never blocks on the kill. The observe is fail-open by design (cursor retained on failure → retried next session), and `runMaterialize` is an idempotent best-effort codex-native side effect (gated by `REMNIC_CODEX_MATERIALIZE`). No data loss, no crash — degraded archival only when the daemon is already too slow to respond. This is the timeout-budget class hardened in ea933589; the values are tuned for real (sub-second) daemon responses.

File: packages/plugin-codex/hooks/hooks.json [hunk 16]
```diff
@@ -8,7 +8,7 @@
             "type": "command",
             "command": "\"${PLUGIN_ROOT}/hooks/bin/remnic-codex-hook.sh\" session-start",
             "commandWindows": "powershell -NoProfile -ExecutionPolicy Bypass -File \"${PLUGIN_ROOT}\\hooks\\bin\\remnic-codex-hook.ps1\" session-start",
-            "timeout": 45000
+            "timeout": 45
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/hooks.json:11]: Won't-fix (fail-open, same timeout-budget class). On a daemon slow enough to time out the 45s full recall, the 20s minimal fallback indeed can't complete before the 45s ceiling — but that path only triggers when the daemon is already unresponsive, and a killed SessionStart hook is fail-open by Codex contract (the session starts without recall context; recall is read-only context injection, never a gate). No data loss, no observe/cursor impact. Real recall round-trips are sub-second; this only degrades when the daemon is pathologically slow, where skipping recall is the correct behavior anyway. Tuning the httpPost budgets further is a preference, not a defect.

File: packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs [hunk 17]
```diff
@@ -26,28 +26,88 @@ const fs = require("fs");
...
+      "http(s):// URL or unset REMNIC_DAEMON_URL to use REMNIC_HOST/REMNIC_PORT.";
+    if (process.env.REMNIC_HOOK_QUIET !== "1") {
+      try { process.stderr.write("[remnic-codex-hook] " + DAEMON_CONFIG_ERROR + "\n"); } catch {}
+    }
+    return null;
+  }
+  try {
+    return new URL(`http://${HOST}:${PORT}`);
+  } catch {
+    return new URL("http://127.0.0.1:4318");
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/bin/remnic-codex-hook.cjs:77]: Won't-fix (scope-expansion of the URL-safety class + documented backward-compat). The invalid-`REMNIC_DAEMON_URL`-disables path (ea933589) is the high-risk case — an explicit full-URL typo silently cross-routing to localhost. The `REMNIC_HOST`/`REMNIC_PORT` path is the legacy/default config: `new URL(\`http://\${HOST}:\${PORT}\`)` in try/catch with a `127.0.0.1:4318` final fallback is the documented, backward-compatible behavior (existing installs and the test harness rely on it). A malformed HOST/PORT (e.g. unbracketed `::1`) is an operator misconfiguration of the legacy pair, not a silent cross-store hazard of the kind that justified disabling on a bad explicit daemon URL — the default `127.0.0.1:4318` is itself the local daemon the operator almost certainly intended. Hardening the HOST/PORT parse the same way is a reasonable follow-up but out of scope for #1571 (Codex host parity); the explicit-URL safety gate covers the fleet/remote case that motivated this PR.

File: packages/plugin-codex/hooks/hooks.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/hooks.json:50]: Won't-fix — same defect class as the cursor thread on this line (PRRT_kwDORJXyws6OYZ8T), already declined there with full evidence: `handleSessionEnd` emits `{continue:true}` FIRST, so Codex never blocks on the kill; the final observe is fail-open by design (cursor retained → retried next session); `runMaterialize` is an idempotent best-effort side effect gated by `REMNIC_CODEX_MATERIALIZE`. No data loss, no crash — only degraded archival when the daemon is already too slow to respond. Declining as a same-class re-raise of the timeout-budget class hardened in ea933589.

File: packages/plugin-codex/hooks/hooks.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/hooks.json:11]: Won't-fix — same defect class as the cursor thread on this line (PRRT_kwDORJXyws6OYZ8U), already declined there: the 65s worst case only materializes when the daemon is already too slow to answer a 45s recall, and a killed SessionStart hook is fail-open by Codex contract (the session starts without read-only recall context). No observe/cursor/data impact. Declining as a same-class re-raise of the timeout-budget class hardened in ea933589.

[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/hooks.json:11]: @/tmp/1619_reply.txt

[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/plugin-codex/hooks/hooks.json:11]: Declining this same class with on-record evidence: SessionStart recall is fail-open by the Codex hooks contract — a slow/absent daemon never blocks the session; the 45s budget only bounds the recall attempt. Bumping the timeout is unnecessary; the same class was already declined with evidence earlier in this PR (see the resolved PreCompact/Stop timeout threads). No code change warranted.
